### PR TITLE
Setting order site from basket site

### PR DESCRIPTION
**Pull request by clintonb**
If a basket is associated with a site, the order created from the basket will have the same site.

ECOM-2587

### CODE REVIEW

File: ecommerce/tests/mixins.py [hunk 1]
```diff
@@ -235,7 +235,21 @@ def create_partner(self, name):
...
+        """ Create a new Site, and link it to the specified Partner. """
+        site = Site.objects.create(domain=domain)
+
+        SiteConfiguration.objects.create(
+            site=site,
+            partner=partner,
+            lms_url_root='https://lms.{}'.format(domain),
+            theme_scss_path='/css/path/',
+            payment_processors='fake_processor'
+        )
```
Comment (cpennington) [inline ecommerce/tests/mixins.py]: Should this use `factory_boy`?

File: ecommerce/tests/mixins.py [hunk 2]
```diff
@@ -244,10 +258,12 @@ def setUp(self):
...
         # Create and configure the default site
-        Site.objects.all().delete()
-        self.site = Site.objects.create(id=1, domain=domain)
+        self.partner = self.create_partner('edx')
 
-        self.partner, _created = Partner.objects.get_or_create(short_code='edx', name='edx')
+        Site.objects.all().delete()
+        self.site = self.create_site(domain, self.partner)
+        self.site.id = 1
+        self.site.save()
```
Comment (cpennington) [inline ecommerce/tests/mixins.py]: Why does this have to change the id of the site?

File: ecommerce/tests/mixins.py (hunk 1 above)
Comment (clintonb) [inline ecommerce/tests/mixins.py]: Yes, that would make this so much simpler.

File: ecommerce/extensions/order/utils.py [hunk 3]
```diff
@@ -42,3 +43,23 @@ def basket_id(self, order_number):
...
+        """
+
+        # Pull the order's site from the basket, if the basket has a site and
+        # a site is not already being explicitly set.
+        if basket.site and 'site'not in extra_order_fields:
+            extra_order_fields['site'] = basket.site
+
+        return super(OrderCreator, self).create_order_model(user, basket, shipping_address, shipping_method,
+                                                            shipping_charge, billing_address, total, order_number,
+                                                            status, **extra_order_fields)
```
Comment (cpennington) [inline ecommerce/extensions/order/utils.py]: Nit: I think we tend to drop arguments down on the next line (when they're going to be multiline).

File: ecommerce/extensions/order/tests/test_utils.py [hunk 4]
```diff
@@ -1,8 +1,15 @@
 """Test Order Utility classes """
 from django.test import TestCase, override_settings
+from oscar.core.loading import get_class
+from oscar.test.factories import create_basket
 from oscar.test.newfactories import BasketFactory
 
-from ecommerce.extensions.order.utils import OrderNumberGenerator
+from ecommerce.extensions.fulfillment.status import ORDER
+from ecommerce.extensions.order.utils import OrderNumberGenerator, OrderCreator
```
Comment (cpennington) [inline ecommerce/extensions/order/tests/test_utils.py]: If we load this using the oscar loader, then if we ever accidentally misconfigure the `OrderCreator`, this test will fail and let us know.

File: ecommerce/tests/mixins.py (hunk 2 above)
Comment (clintonb) [inline ecommerce/tests/mixins.py]: The default site ID (defined by the setting `SITE_ID`) is 1. I want to make sure we have the default site set properly. I'm going to update this to use `factory_boy` as well.

Comment (cpennington) [inline ecommerce/tests/mixins.py]: Shouldn't we then use `SITE_ID`, rather than hardcoding 1?

Comment (clintonb) [inline ecommerce/tests/mixins.py]: (facepalm) Yes.

File: ecommerce/extensions/order/utils.py [hunk 5]
```diff
@@ -42,3 +43,23 @@ def basket_id(self, order_number):
...
+        """
+        Create an order model.
+
+        This override ensures the order's site is set to that of the basket. If the basket has no site, the default
+        site is used. The site value can be overridden by setting the `site` kwarg.
+        """
+
+        # Pull the order's site from the basket, if the basket has a site and
+        # a site is not already being explicitly set.
+        if basket.site and 'site'not in extra_order_fields:
```
Comment (jimabramson) [inline ecommerce/extensions/order/utils.py]: `'site' not in`

Comment (jimabramson) [inline ecommerce/extensions/order/utils.py]: why wouldn't a basket be associated with a site?

File: ecommerce/extensions/order/tests/test_utils.py [hunk 6]
```diff
@@ -31,3 +38,51 @@ def test_order_number_generation(self):
...
+    def setUp(self):
+        super(OrderCreatorTests, self).setUp()
+        self.user = self.create_user()
+
+    def create_order_model(self, basket):
+        """ Call the create_order_model method to create an Order from the given Basket. """
+        shipping_method = NoShippingRequired()
+        shipping_charge = shipping_method.calculate(basket)
+        total = OrderTotalCalculator().calculate(basket, shipping_charge)
+        return self.order_creator.create_order_model(self.user, basket, None, NoShippingRequired(), shipping_charge,
```
Comment (jimabramson) [inline ecommerce/extensions/order/tests/test_utils.py]: you already have a `shipping_method`, why not just pass that?

File: ecommerce/extensions/order/utils.py (hunk 5 above)
Comment (clintonb) [inline ecommerce/extensions/order/utils.py]: Pre-multi-tenancy baskets have no site. Until this field is made non-nullable, we cannot assume a site exists.

File: ecommerce/extensions/order/tests/test_utils.py (hunk 6 above)
Comment (clintonb) [inline ecommerce/extensions/order/tests/test_utils.py]: Fixed.

File: ecommerce/extensions/order/utils.py (hunk 5 above)
Comment (clintonb) [inline ecommerce/extensions/order/utils.py]: Fixed.

**Comment by jimabramson** [comment]
BTW, +1 for factory boy

File: ecommerce/extensions/order/utils.py (hunk 5 above)
Comment (jimabramson) [inline ecommerce/extensions/order/utils.py]: understood.

File: ecommerce/extensions/order/utils.py (hunk 3 above)
Comment (clintonb) [inline ecommerce/extensions/order/utils.py]: Updated.

File: ecommerce/extensions/order/tests/test_utils.py (hunk 4 above)
Comment (clintonb) [inline ecommerce/extensions/order/tests/test_utils.py]: Updated.

**Comment by clintonb** [comment]
Review and merge #421 first.

File: ecommerce/extensions/order/tests/test_utils.py [hunk 7]
```diff
@@ -34,3 +40,51 @@ def test_order_number_generation(self):
...
+        """ Call the create_order_model method to create an Order from the given Basket. """
+        shipping_method = NoShippingRequired()
+        shipping_charge = shipping_method.calculate(basket)
+        total = OrderTotalCalculator().calculate(basket, shipping_charge)
+        return self.order_creator.create_order_model(self.user, basket, None, shipping_method, shipping_charge,
+                                                     None, total, basket.order_number, ORDER.OPEN)
+
+    def create_basket(self, site):
+        """ Returns a new Basket with the specified Site. """
+        basket = create_basket()
```
Comment (rlucioni) [inline ecommerce/extensions/order/tests/test_utils.py]: It took me a few takes to understand what was going on here. Maybe import `create_basket` as `oscar_create_basket`?

**Comment by rlucioni** [comment]
:+1:

File: ecommerce/extensions/order/tests/test_utils.py (hunk 7 above)
Comment (clintonb) [inline ecommerce/extensions/order/tests/test_utils.py]: Done.

**Pull request merged**